src/Interpreters/GraceHashJoin.cpp

```diff
@@ -401,9 +403,14 @@ void GraceHashJoin::checkTypesOfKeys(const Block & block) const
 
 void GraceHashJoin::initialize(const Block & sample_block)
 {
+    std::lock_guard lock(hash_join_mutex);
```

Comment:
I think this fix is not right since this function is exeucted in single thread, and also, if this is the problem, maybe it should be found early in many days, right?